File: python-webrtc/cpp/src/interfaces/rtc_rtp_transceiver.cpp
//
// Copyright 2022 Il`ya (Marshal) <https://github.com/MarshalX>. All rights reserved.
//
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE.md file in the root of the project.
//

#include "rtc_rtp_transceiver.h"
#include "../exceptions.h"

namespace python_webrtc {

  RTCRtpTransceiver::RTCRtpTransceiver(PeerConnectionFactory *factory,
                                       rtc::scoped_refptr<webrtc::RtpTransceiverInterface> transceiver) : _factory(
      factory), _transceiver(std::move(transceiver)) {}

  RTCRtpTransceiver::~RTCRtpTransceiver() {
    _factory == nullptr;
    holder()->Release(this);
  }

  void RTCRtpTransceiver::Init(pybind11::module &m) {
    pybind11::class_<RTCRtpTransceiver>(m, "RTCRtpTransceiver")
        .def_property_readonly("mid", &RTCRtpTransceiver::GetMid)
        .def_property_readonly("sender", &RTCRtpTransceiver::GetSender, pybind11::return_value_policy::reference)
        .def_property_readonly("receiver", &RTCRtpTransceiver::GetReceiver, pybind11::return_value_policy::reference)
        .def_property_readonly("stopped", &RTCRtpTransceiver::GetStopped)
        .def_property("direction", &RTCRtpTransceiver::GetDirection, &RTCRtpTransceiver::SetDirection)
        .def_property_readonly("currentDirection", &RTCRtpTransceiver::GetCurrentDirection)
            // set codec pref
        .def("stop", &RTCRtpTransceiver::Stop);
  }

  InstanceHolder<RTCRtpTransceiver *, rtc::scoped_refptr<webrtc::RtpTransceiverInterface>, PeerConnectionFactory *> *
  RTCRtpTransceiver::holder() {
    static auto holder = new InstanceHolder<
        RTCRtpTransceiver *, rtc::scoped_refptr<webrtc::RtpTransceiverInterface>, PeerConnectionFactory *
    >(RTCRtpTransceiver::Create);
    return holder;
  }

  RTCRtpTransceiver *RTCRtpTransceiver::Create(PeerConnectionFactory *factory,
                                               rtc::scoped_refptr<webrtc::RtpTransceiverInterface> transceiver) {
    // who caring about freeing memory?
    return new RTCRtpTransceiver(factory, std::move(transceiver));
  }

  std::optional<std::string> RTCRtpTransceiver::GetMid() {
    if (_transceiver->mid()) {
      return _transceiver->mid().value();
    }

    return {};
  }

  RTCRtpSender *RTCRtpTransceiver::GetSender() {
    return RTCRtpSender::holder()->GetOrCreate(_factory, _transceiver->sender());
  }

  RTCRtpReceiver *RTCRtpTransceiver::GetReceiver() {
    return RTCRtpReceiver::holder()->GetOrCreate(_factory, _transceiver->receiver());
  }

  bool RTCRtpTransceiver::GetStopped() {
    // TODO Deprecated: This feature is no longer recommended.
    return _transceiver->stopped();
  }

  webrtc::RtpTransceiverDirection RTCRtpTransceiver::GetDirection() {
    return _transceiver->direction();
  }

  void RTCRtpTransceiver::SetDirection(webrtc::RtpTransceiverDirection direction) {
    auto result = _transceiver->SetDirectionWithError(direction);
    if (!result.ok()) {
      throw wrapRTCError(result);
    }
  }

  std::optional<webrtc::RtpTransceiverDirection> RTCRtpTransceiver::GetCurrentDirection() {
    if (_transceiver->current_direction()) {
      return _transceiver->current_direction().value();
    }

    return {};
  }

  void RTCRtpTransceiver::Stop() {
    auto result = _transceiver->StopStandard();
    if (!result.ok()) {
      throw wrapRTCError(result);
    }
  }

} // namespace python_webrtc

File: python-webrtc/python/webrtc/interfaces/rtc_rtp_transceiver.py
#
#  Copyright 2022 Il`ya (Marshal) <https://github.com/MarshalX>. All rights reserved.
#
#  Use of this source code is governed by a BSD-style license
#  that can be found in the LICENSE.md file in the root of the project.
#

from typing import TYPE_CHECKING, Optional

from webrtc import wrtc, WebRTCObject

if TYPE_CHECKING:
    import webrtc


class RTCRtpTransceiver(WebRTCObject):
    """The WebRTC interface :obj:`webrtc.RTCRtpTransceiver` describes a permanent pairing of
    an :obj:`webrtc.RTCRtpSender` and an :obj:`webrtc.RTCRtpReceiver`, along with some shared state.
    """

    _class = wrtc.RTCRtpTransceiver

    @property
    def mid(self) -> Optional[str]:
        """A :obj:`str` which uniquely identifies the pairing of source and destination of the transceiver's stream.
        Its value is taken from the media ID of the SDP m-line. This value is :obj:`None` if negotiation
        has not completed."""
        return self._native_obj.mid

    @property
    def receiver(self) -> 'webrtc.RTCRtpReceiver':
        """A :obj:`webrtc.RTCRtpReceiver` object which is responsible for receiving and decoding incoming media data
        whose media ID is the same as the current value of :attr:`mid`."""
        from webrtc import RTCRtpReceiver

        return RTCRtpReceiver._wrap(self._native_obj.receiver)

    @property
    def sender(self) -> 'webrtc.RTCRtpSender':
        """A :obj:`webrtc.RTCRtpSender` object used to encode and send media whose media ID matches
        the current value of :attr:`mid`."""
        from webrtc import RTCRtpSender

        return RTCRtpSender._wrap(self._native_obj.sender)

    @property
    def stopped(self) -> bool:
        """A :obj:`bool` value which is :obj:`True` if the transceiver's :attr:`sender` will no longer send data,
        and its :attr:`receiver` will no longer receive data. If either or both are still at work,
        the result is :obj:`False`.

        Warning:
            Deprecated: This feature is no longer recommended.
        """
        return self._native_obj.stopped

    @property
    def direction(self) -> 'webrtc.TransceiverDirection':
        """A member of :obj:`webrtc.TransceiverDirection` enum, indicating the transceiver's preferred direction.

        Note:
            The transceiver's current direction is indicated by the :attr:`currentDirection` property.
        """
        return self._native_obj.direction

    @direction.setter
    def direction(self, new_direction: 'webrtc.TransceiverDirection'):
        self._native_obj.direction = new_direction

    @property
    def current_direction(self) -> Optional['webrtc.TransceiverDirection']:
        """A member of :obj:`webrtc.TransceiverDirection` enum, indicating
        the current directionality of the transceiver."""
        return self._native_obj.currentDirection

    def stop(self) -> None:
        """Permanently stops the transceiver by stopping both the associated :obj:`webrtc.RTCRtpSender`
        and :obj:`webrtc.RTCRtpReceiver`.

        Note:
            The :attr:`stopped` property was provided to return :obj:`True` if the connection is stopped.
            That property has been deprecated and will be removed at some point. Instead, check the value
            of :attr:`currentDirection`. If it's :obj:`webrtc.TransceiverDirection.stopped`, the transceiver
            has been stopped.
        """
        self._native_obj.stop()

    #: Alias for :attr:`current_direction`
    currentDirection = current_direction
